### PR TITLE
CGMES: Fix support for ACLineSegments contained in cim:Line containers

### PR DESCRIPTION
**Pull request by zamarrenolm**
**Please check if the PR fulfills these requirements**

- [X] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

**What kind of change does this PR introduce?**

Bug fix

**What is the current behavior?**

Some configurations of sets of `ACLineSegments` and `ConnectivityNodes` or `TopologicalNodes` inside CGMES `Line` containers were not supported.

**What is the new behavior (if this is a feature change)?**
The existing strategy for creating fictitious IIDM Voltage Level containers for `cim:Line` containers has been fixed to support the observed configurations.
Creation of the fictitious containers has been moved before converting the nodes.
Explicit creation of new nodes is now not required in the code that prepares the fictitious containers.
In some situations it may not be possible to determine a nominal voltage reference for the fictitious IIDM voltage level. An error is logged (there is a unit test where this happens: `CgmesConformity1ModifiedConversionTest::miniNodeBreakerMissingSubstationRegion`).

**Does this PR introduce a breaking change or deprecate an API?**
- [ ] Yes
- [X] No

**Other information**:

The problem was observed trying to import National Grid public CIM data from https://www.nationalgrid.co.uk/our-network/energy-data-hub/common-information-model. Specifically the "Iron Acton" files from South West.

A future improvement may be to create a fictitious voltage level for each `cim:Line` instead of a voltage level for each `node`. 

Also, we should not rely only on ACLSs to determine the creation of fictitious containers. The association ACLS -> Line is optional. Line containers should be obtained from nodes. Only Line containers containing nodes should be mapped to fictitious voltage levels.

### CODE REVIEW

**Comment by zamarrenolm** [comment]
A configuration still not supported exists in another case from the same National Grid dataset. The "Indian Queens" network contains at least a `Switch` that has one end at a connectivity node inside a voltage level inside a substation and the other end at a connectivity node inside a `cim:Line` container. Currently, we only support ACLSs with this kind of connection.

**Comment by zamarrenolm** [comment]
Many examples in the East and West Midland dataset contain references to containers that are not defined in the CGMES in the same input file. These missing references can be automatically created during import using the pre-processor that automatically creates missing containers in the input files. The pre-processor generates an output EQ file that contains basic CGMES definitions for them. This behaviour can be configured following this example:

```
import-export-parameters-default-value:
  iidm.import.cgmes.pre-processors: "createMissingContainers"
  iidm.import.cgmes.fixes-for-missing-containers-folder: "/output/folder/for/completions"
```

**Comment by agile-scrum-xp** [comment]
Thanks Luma Zamarreño. 

I am running the following code:
```
private static Network loadNetworkFromCIM() {

        File fileIronActon = new File(SldCustomNodeTutorial.class.getResource("/iron_acton_swest_sgp.zip").getPath());

        Network networkIronActon = Network.read(fileIronActon.toString());

        System.out.println("ID of Rugeley network: " + networkIronActon.getId());
        System.out.println("Rugeleysubstation count :" + networkIronActon.getSubstationCount());
        System.out.println("Rugeleysubstation count :" + networkIronActon.getVoltageLevels().iterator().next());

        return networkIronActon;
    }
```

but still getting the error:

```
[main] WARN com.powsybl.cgmes.conversion.Conversion - Fictitious Voltage Level INM_ID_FFF0DB7F-D02A-4053-9772-9650D40B73F8_CN1_VL created for Line container 264F6C38-1B3A-4211-A287-BB67FD5378EB node LIN:110641-[167838460]-112014
Exception in thread "main" com.powsybl.cgmes.conversion.ConversionException: Boundary must be at one end of the line
	at com.powsybl.cgmes.conversion.elements.ACLineSegmentConversion.convertAtBoundary(ACLineSegmentConversion.java:57)
	at com.powsybl.cgmes.conversion.Conversion.convertEquipmentAtBoundaryNode(Conversion.java:686)
	at com.powsybl.cgmes.conversion.Conversion.lambda$convert$19(Conversion.java:211)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at com.powsybl.cgmes.conversion.Conversion.convert(Conversion.java:211)
	at com.powsybl.cgmes.conversion.CgmesImport.importData1(CgmesImport.java:169)
	at com.powsybl.cgmes.conversion.CgmesImport.importData(CgmesImport.java:163)
	at com.powsybl.iidm.network.Network.read(Network.java:127)
	at com.powsybl.iidm.network.Network.read(Network.java:134)
	at com.powsybl.iidm.network.Network.read(Network.java:148)
	at com.powsybl.iidm.network.Network.read(Network.java:162)
	at com.powsybl.iidm.network.Network.read(Network.java:177)
	at com.powsybl.iidm.network.Network.read(Network.java:187)
	at com.powsybl.tutorials.sld.customnode.SldCustomNodeTutorial.loadNetworkFromCIM(SldCustomNodeTutorial.java:160)
	at com.powsybl.tutorials.sld.customnode.SldCustomNodeTutorial.main(SldCustomNodeTutorial.java:130)
```

I am not sure if I am doing something wrong. I am using powsybl-tutorials-main and these are my steps:
1. Downloaded the latest version from powsybl-core
2. compiled and installed the downloaded version `mvn install -DskipTests`
3. Now, the local mvn repository has 
       ```
 <groupId>com.powsybl</groupId>
        <artifactId>powsybl-parent</artifactId>
        <version>18</version>
```
4. I updated the pom.xml in powsybl-tutorials-main earlier it was (8)
```
 <parent>
        <groupId>com.powsybl</groupId>
        <artifactId>powsybl-parent</artifactId>
        <version>18</version>
        <relativePath/>
    </parent>
```
5. The pom.xml still has 
```
<dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>com.powsybl</groupId>
                <artifactId>powsybl-dependencies</artifactId>
                <version>2024.0.3</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
        </dependencies>
    </dependencyManagement>
```

Not sure if I am doing wrong and the system is picking the old jar file/s. I am not using iTools but Java to load design.

Thanks for all the help.

**Comment by zamarrenolm** [comment]
> Thanks Luma Zamarreño.
> 
> I am running the following code:
> 
> ```
> private static Network loadNetworkFromCIM() {
> 
>         File fileIronActon = new File(SldCustomNodeTutorial.class.getResource("/iron_acton_swest_sgp.zip").getPath());
> 
>         Network networkIronActon = Network.read(fileIronActon.toString());
> 
>         System.out.println("ID of Rugeley network: " + networkIronActon.getId());
>         System.out.println("Rugeleysubstation count :" + networkIronActon.getSubstationCount());
>         System.out.println("Rugeleysubstation count :" + networkIronActon.getVoltageLevels().iterator().next());
> 
>         return networkIronActon;
>     }
> ```
> 
> but still getting the error:
> 
> ```
> [main] WARN com.powsybl.cgmes.conversion.Conversion - Fictitious Voltage Level INM_ID_FFF0DB7F-D02A-4053-9772-9650D40B73F8_CN1_VL created for Line container 264F6C38-1B3A-4211-A287-BB67FD5378EB node LIN:110641-[167838460]-112014
> Exception in thread "main" com.powsybl.cgmes.conversion.ConversionException: Boundary must be at one end of the line
> 	at com.powsybl.cgmes.conversion.elements.ACLineSegmentConversion.convertAtBoundary(ACLineSegmentConversion.java:57)
> 	at com.powsybl.cgmes.conversion.Conversion.convertEquipmentAtBoundaryNode(Conversion.java:686)
> 	at com.powsybl.cgmes.conversion.Conversion.lambda$convert$19(Conversion.java:211)
> 	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
> 	at com.powsybl.cgmes.conversion.Conversion.convert(Conversion.java:211)
> 	at com.powsybl.cgmes.conversion.CgmesImport.importData1(CgmesImport.java:169)
> 	at com.powsybl.cgmes.conversion.CgmesImport.importData(CgmesImport.java:163)
> 	at com.powsybl.iidm.network.Network.read(Network.java:127)
> 	at com.powsybl.iidm.network.Network.read(Network.java:134)
> 	at com.powsybl.iidm.network.Network.read(Network.java:148)
> 	at com.powsybl.iidm.network.Network.read(Network.java:162)
> 	at com.powsybl.iidm.network.Network.read(Network.java:177)
> 	at com.powsybl.iidm.network.Network.read(Network.java:187)
> 	at com.powsybl.tutorials.sld.customnode.SldCustomNodeTutorial.loadNetworkFromCIM(SldCustomNodeTutorial.java:160)
> 	at com.powsybl.tutorials.sld.customnode.SldCustomNodeTutorial.main(SldCustomNodeTutorial.java:130)
> ```
> 
> I am not sure if I am doing something wrong. I am using powsybl-tutorials-main and these are my steps:
> 
> 1. Downloaded the latest version from powsybl-core
> 2. compiled and installed the downloaded version `mvn install -DskipTests`
> 3. Now, the local mvn repository has
>    ```
>    com.powsybl
>    powsybl-parent
>    18
> 
> ```
> 4. I updated the pom.xml in powsybl-tutorials-main earlier it was (8)
> ```
> 
> com.powsybl powsybl-parent 18 ``` 5. The pom.xml still has ``` com.powsybl powsybl-dependencies 2024.0.3 pom import ```
> Not sure if I am doing wrong and the system is picking the old jar file/s. I am not using iTools but Java to load design.
> 
> Thanks for all the help.

If you want to use this fix you have to obtain the code in the branch associated with this PR (it still not merged in the latest version). You can do it using the git command line with:

```
git clone -b  cgmes_import_fix_support_acls_in_cim_line_containers https://github.com/powsybl/powsybl-core.git
```

After the checkout, your powsybl-core/pom.xml should contain the lines:

```xml
    <artifactId>powsybl-core</artifactId>
    <version>6.3.0-SNAPSHOT</version>
```

Then you can build all the modules, the built version will contain the fix. You have to update the dependencies in your project to use the artifacts produced (depend on powsybl-core version 6.3.0-SNAPSHOT).

If you want to avoid other errors during import (existing references to non-defined containers) remember to use the configuration provided in my previous comment.

**Review by marqueslanauja — APPROVED** [review]

File: cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/Conversion.java [hunk 1]
```diff
@@ -546,37 +567,30 @@ private void convertACLineSegmentsToLines(Context context, Set<String> delayedBo
...
-                    delayedBoundaryNodes.add(node);
+                if (vlRef != null) {
+                    createLineContainerFictitiousVoltageLevels(context, lineContainerId, vlRef, line);
                 } else {
-                    c.convert();
+                    LOG.error("No fictitious Voltage Level created for Line container {}. No voltage level reference could be found", lineContainerId);
                 }
             }
         }
     }
```
[the remark this answers is not in the document]
Comment (zamarrenolm) [reply cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/Conversion.java:583]: added a comment

**Comment by zamarrenolm** [comment]
Added a specific unit that reproduces the configuration where the problem was observed

**Pull request merged**